<launch>
  <!-- Filtration parameters -->
  <arg name="UAV_NAME" default="$(env UAV_NAME)" />

  <arg name="lidar3d/republish" default="true" />
  <arg name="rplidar_republish" default="false" />

  <arg name="lidar3d/topic_in" default="os_cloud_nodelet/points" />
  <arg name="lidar3d/keep_organized" default="true" />
  <arg name="lidar3d/range_clip/use" default="true" />
  <arg name="lidar3d/range_clip/min" default="0.45" />
  <arg name="lidar3d/range_clip/max" default="20.0" />
  <arg name="lidar3d/ground_removal/use" default="false" />
  <arg name="lidar3d/downsampling/dynamic_row_selection" default="false" />
  <arg name="lidar3d/downsampling/row_step" default="1" doc="dynamic==false: will keep every n-th row; dynamic==true: number of iterations" />
  <arg name="lidar3d/downsampling/col_step" default="1" doc="will keep every n-th column" />
  <arg name="lidar3d/filter_intensity/use" default="false" />
  <arg name="lidar3d/filter_intensity/threshold" default="100" />
  <arg name="lidar3d/filter_intensity/range" default="10.0" />

  <arg name="depth/custom_config" default="" />
  <!-- <arg name="depth/custom_config" default="$(find mrs_pcl_tools)/config/two_realsenses.yaml" /> -->

  <arg name="nodelet" default="standalone" />
  <arg name="nodelet_manager" default="" />
  <arg name="launch_prefix" default=""/>

  <group ns="$(arg UAV_NAME)">
    <node pkg="nodelet" type="nodelet" name="pcl_filtration" args="$(arg nodelet) mrs_pcl_tools/PCLFiltration $(arg nodelet_manager)" output="screen" launch-prefix="$(arg launch_prefix)" respawn="true" >

      <!-- Parameters -->
      <param name="uav_name" type="string" value="$(arg UAV_NAME)" />

      <param name="lidar3d/republish" type="bool" value="$(arg lidar3d/republish)" />
      <param name="lidar3d/keep_organized" type="bool" value="$(arg lidar3d/keep_organized)" />
      <param name="lidar3d/range_clip/use" type="bool" value="$(arg lidar3d/range_clip/use)" />
      <param name="lidar3d/range_clip/min" type="double" value="$(arg lidar3d/range_clip/min)" />
      <param name="lidar3d/range_clip/max" type="double" value="$(arg lidar3d/range_clip/max)" />
      <param name="lidar3d/ground_removal/use" type="bool" value="$(arg lidar3d/ground_removal/use)" />
      <param name="lidar3d/ground_removal/range/use" type="bool" value="true" />
      <param name="lidar3d/ground_removal/static_frame_id" type="string" value="gps_origin" />
      <param name="lidar3d/ground_removal/max_precrop_height" type="double" value="2.0" />
      <param name="lidar3d/ground_removal/ransac/max_inlier_distance" type="double" value="0.5" />
      <param name="lidar3d/ground_removal/offset" type="double" value="1.0" />
      <param name="lidar3d/downsampling/dynamic_row_selection" type="bool" value="$(arg lidar3d/downsampling/dynamic_row_selection)" />
      <param name="lidar3d/downsampling/row_step" type="int" value="$(arg lidar3d/downsampling/row_step)" />
      <param name="lidar3d/downsampling/col_step" type="int" value="$(arg lidar3d/downsampling/col_step)" />
      <param name="lidar3d/filter/intensity/use" type="bool" value="$(arg lidar3d/filter_intensity/use)" />
      <param name="lidar3d/filter/intensity/threshold" type="int" value="$(arg lidar3d/filter_intensity/threshold)" />
      <param name="lidar3d/filter/intensity/range" type="double" value="$(arg lidar3d/filter_intensity/range)" />

      <!-- Load depth camera parameters from a config file -->
      <rosparam if="$(eval not arg('depth/custom_config') == '')" file="$(arg depth/custom_config)" command="load" />

      <param name="rplidar/republish" type="bool" value="$(arg rplidar_republish)" />
      <param name="rplidar/voxel_resolution" type="double" value="0.05" />

      <!-- Subscribers -->
      <remap from="~lidar3d_in" to="$(arg lidar3d/topic_in)" />
      <remap from="~rplidar_in" to="rplidar/scan_raw" />
      <remap from="~rangefinder_in" to="mavros/distance_sensor/garmin" />

      <!-- Publishers -->
      <remap from="~lidar3d_out" to="$(arg lidar3d/topic_in)_processed" />
      <remap from="~lidar3d_over_max_range_out" to="$(arg lidar3d/topic_in)_over_max_range" />
      <remap from="~rplidar_out" to="rplidar/scan_processed" />

    </node>

  </group>

</launch>
